packages/commons/src/context/context.ts

```diff
@@ -25,21 +27,38 @@ export function fromAppContext(ctx: AppContext): WithLogger<AppContext> {
   return { ...ctx, logger: logger({ ...ctx }) };
 }
 
+const makeApiProblem = makeApiProblemBuilder({});
+
 export const contextMiddleware =
   (
     serviceName: string,
-    overrideCorrelationId: boolean = false
+    readCorrelationIdFromHeader: boolean = true
   ): ZodiosRouterContextRequestHandler<ExpressContext> =>
-  (req, _res, next): void => {
-    const correlationId = overrideCorrelationId
-      ? uuidv4()
-      : readCorrelationIdHeader(req) ?? uuidv4();
-
-    // eslint-disable-next-line functional/immutable-data
-    req.ctx = {
-      serviceName,
-      correlationId,
-    } as AppContext;
-
-    next();
+  async (req, res, next): Promise<unknown> => {
+    const setCtx = (correlationId: string): void => {
+      // eslint-disable-next-line functional/immutable-data
+      req.ctx = {
+        serviceName,
+        correlationId,
+      } as AppContext;
+    };
+
+    if (readCorrelationIdFromHeader) {
+      const correlationIdHeader = parseCorrelationIdHeader(req);
+
+      if (!correlationIdHeader) {
+        const problem = makeApiProblem(
+          missingHeader("X-Correlation-Id"),
+          () => constants.HTTP_STATUS_BAD_REQUEST,
+          genericLogger
+        );
+        return res.status(problem.status).send(problem);
+      }
+
+      setCtx(correlationIdHeader || uuidv4());
```

Comment:
how works this check? 
all falsy values for `correlationIdHeader` are handled by previous if statement, so it should never using `uuidv4()` I think. The only case where we add random uuid is when `readCorrelationIdFromHeader` is false
```suggestion
      setCtx(correlationIdHeader);
```